src/signatures/secp256k1.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use core::convert::TryInto;
+use k256::{
+    elliptic_curve::{
+        ecdh,
+        generic_array::{typenum::Unsigned, GenericArray},
+        sec1::{Coordinates, EncodedPoint, ToEncodedPoint, UntaggedPointSize},
+    },
+    Secp256k1,
+};
+use signature::{Error as SigError, Signer, Verifier};
+
+/// A byte representation of an Secp256k1 field element.
+pub type FieldBytes = k256::FieldBytes;
+
+/// Secp256k1 Shared Secret - the result of a Diffie-Hellman key exchange.
+///
+/// Each party computes this from an (ephemeral) [SecretKey] and their counterparty's [PublicKey].
+pub type SharedSecret = k256::ecdh::SharedSecret;
+
+/// Secp256k1 Public Key
+pub struct PublicKey(k256::PublicKey);
+
+// algorithm identifier used for errors
+const NAME: &str = "Secp256k1";
+
+impl PublicKey {
+    /// Load a [PublicKey] from a slice of bytes.
+    pub fn from_bytes(bytes: &[u8]) -> crate::Result<Self> {
+        assert_buffer_eq!(bytes.len(), UntaggedPointSize::<Secp256k1>::USIZE, "bytes");
+
+        EncodedPoint::from_untagged_bytes(bytes.into())
+            .decode()
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "bytes",
+                to: "Secp256k1 Public Key",
+            })
+    }
+
+    /// Load a [PublicKey] from big-endian serialized coordinates.
+    pub fn from_coord(x: &[u8], y: &[u8]) -> crate::Result<Self> {
+        let x: &FieldBytes = x.try_into().map_err(|_| crate::Error::ConvertError {
+            from: "bytes",
+            to: "Secp256k1 Point (x)",
+        })?;
+
+        let y: &FieldBytes = y.try_into().map_err(|_| crate::Error::ConvertError {
+            from: "bytes",
+            to: "Secp256k1 Point (y)",
+        })?;
+
+        EncodedPoint::from_affine_coordinates(x, y, false)
+            .try_into()
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "Coordinates",
+                to: "Secp256k1 Public Key",
+            })
+    }
+
+    /// Return the [PublicKey] as an array of bytes.
+    pub fn to_bytes(&self) -> GenericArray<u8, UntaggedPointSize<Secp256k1>> {
```

Comment:
Why not just `[u8; PUBLIC_KEY_LEN]`?

Reply:
Rust now supports generic array legnths natively as in `fn f<const N: usize>() -> [u8; N]`. Shall we move closer to using this feature rather `GenericArray`?